In [1]:
#import dependencies
import pandas as pd

In [3]:
# read in data from CSV and create a dataframe

filepath = "2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv"

file = pd.read_csv(filepath)

df = pd.DataFrame(file)

df.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [4]:
# create a copy dataframe to work with

sales_df = df.copy()

### This assignment states that there seems to be something wrong with a previously calculated average order value so the first step is to sort the data by order amount and look for any orders that seem unusual.

In [13]:
# sort the data by order_amount from greatest to least and inspect the results

sorted_df = sales_df.sort_values("order_amount",axis=0,ascending=False)

sorted_df.head(50)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
2153      2154       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
15          16       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
691        692       78      878        154350            6          debit   
2492      2493       78      834        102900            4          debit   
3724      3725       78      766         77175            3    credit_card   
4420      4421       78      969         77175            3          debit   
4192      4193       78      787         77175            3    credit_card   
3403      3404       78      928         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2564      2565       78      915         77175            3          debit   
4715      4716       78      818         77175            3          debit   
1259      1260       78      775         77175            3    credit_card   
2906      2907       78      817         77175            3          debit   
3705      3706       78      828         51450            2    credit_card   
3101      3102       78      855         51450            2    credit_card   
4412      4413       78      756         51450            2          debit   
3167      3168       78      927         51450            2           cash   
490        491       78      936         51450            2          debit   
4079      4080       78      946         51450            2           cash   
1529      1530       78      810         51450            2           cash   
4311      4312       78      960         51450            2          debit   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
617        618       78      760         51450            2           cash   
2512      2513       78      935         51450            2          debit   
511        512       78      967         51450            2           cash   
2452      2453       78      709         51450            2           cash   
493        494       78      983         51450            2           cash   
2495      2496       78      707         51450            2           cash   
4040      4041       78      852         25725            1           cash   
4918      4919       78      823         25725            1           cash   
1056      1057       78      800         25725            1          debit   
2922      2923       78      740         25725            1          debit   
2270      2271       78      855         25725            1    credit_card   
1193      1194       78      944         25725            1          debit   

               created_at  

In [14]:
# look at all orders by id 607

df_607 = sales_df.loc[sales_df['user_id'] == 607]

df_607

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  
15    2017-03-07 4:00:00  
60    2017-03-04 4:00:00  
520   2017-03-02 4:00:00  
1104  2017-03-24 4:00:00  
1362  2017-03-15 4:00:00  
1436  2017-03-11 4:00:00  
1562  2017-03-19 4:00:00  
1602  2017-03-17 4:00:00  
2153  2017-03-12 4:00:00  
2297  2017-03-07 4:00:00  
2835  2017-03-28 4:00:00  
2969  2017-03-28 4:00:00  
3332  2017-03-24 4:00:00  
4056  2017-03-28 4:00:00  
4646  2017-03-02 4:00:00  
4868  2017-03-22 4:00:00  
4882  2017-03-25 4:00:00

In [15]:
df_42 = sales_df.loc[sales_df['shop_id'] == 42]
df_42

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
40          41       42      793           352            1    credit_card   
60          61       42      607        704000         2000    credit_card   
308        309       42      770           352            1    credit_card   
409        410       42      904           704            2    credit_card   
520        521       42      607        704000         2000    credit_card   
834        835       42      792           352            1           cash   
835        836       42      819           704            2           cash   
938        939       42      808          1056            3    credit_card   
979        980       42      744           352            1          debit   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1364      1365       42      797          1760            5           cash   
1367      1368       42      926          1408            4           cash   
1436      1437       42      607        704000         2000    credit_card   
1471      1472       42      907          1408            4          debit   
1512      1513       42      946           352            1          debit   
1520      1521       42      756           704            2          debit   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
1911      1912       42      739           704            2           cash   
1929      1930       42      770           352            1    credit_card   
2003      2004       42      934           704            2           cash   
2018      2019       42      739           352            1          debit   
2053      2054       42      951           352            1          debit   
2153      2154       42      607        704000         2000    credit_card   
2273      2274       42      747           704            2          debit   
2297      2298       42      607        704000         2000    credit_card   
2491      2492       42      868           704            2          debit   
2609      2610       42      868           704            2          debit   
2766      2767       42      970           704            2    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
2987      2988       42      819          1056            3           cash   
3332      3333       42      607        704000         2000    credit_card   
3513      3514       42      726          1056            3          debit   
3651      3652       42      830           352            1    credit_card   
3697      3698       42      839           352            1          debit   
3903      3904       42      975           352            1          debit   
3998      3999       42      886           352            1          debit   
4056      4057       42      607        704000         2000    credit_card   
4231      4232       42      962           352            1           cash   
4294      4295       42      859           704            2           cash   
4326      4327       42      788           704            2          debit   
4421      4422       42      736           704            2    credit_card   
4625      4626       42      809           352            1    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4745      4746       42      872           352            1          debit   
4767      4768       42      720           704            2    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42     

### Right away it becomes clear that there are some transactions that would throw off the calculation of average order value, and they fall into two types. The first type is a series of transactions by a single user id, 607, all done by credit card at store 42. Each transaction by id 607 is for 2,000 items at a cost of 704,000, which averages to 352/item. Other smaller transactions occurred at store 42 with similar prices per item so the issue does not seem to be the store. Ideally, I would need to find out more information about user id 607 in order to handle it properly but without being able to do that treating 607 as an outlier and excluding all transactions by user id 607 from calculations is the best option. 

In [18]:
# drop all transactions by user id 607 from the dataframe

step_one = sorted_df[sorted_df.user_id != 607]
step_one.head(50)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
691        692       78      878        154350            6          debit   
2492      2493       78      834        102900            4          debit   
3724      3725       78      766         77175            3    credit_card   
4420      4421       78      969         77175            3          debit   
4192      4193       78      787         77175            3    credit_card   
3403      3404       78      928         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2564      2565       78      915         77175            3          debit   
4715      4716       78      818         77175            3          debit   
1259      1260       78      775         77175            3    credit_card   
2906      2907       78      817         77175            3          debit   
3705      3706       78      828         51450            2    credit_card   
3101      3102       78      855         51450            2    credit_card   
4412      4413       78      756         51450            2          debit   
3167      3168       78      927         51450            2           cash   
490        491       78      936         51450            2          debit   
4079      4080       78      946         51450            2           cash   
1529      1530       78      810         51450            2           cash   
4311      4312       78      960         51450            2          debit   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
617        618       78      760         51450            2           cash   
2512      2513       78      935         51450            2          debit   
511        512       78      967         51450            2           cash   
2452      2453       78      709         51450            2           cash   
493        494       78      983         51450            2           cash   
2495      2496       78      707         51450            2           cash   
4040      4041       78      852         25725            1           cash   
4918      4919       78      823         25725            1           cash   
1056      1057       78      800         25725            1          debit   
2922      2923       78      740         25725            1          debit   
2270      2271       78      855         25725            1    credit_card   
1193      1194       78      944         25725            1          debit   
1452      1453       78      812         25725            1    credit_card   
3780      3781       78      889         25725            1           cash   
4505      4506       78      866         25725            1          debit   
2773      2774       78      890         25725            1           cash   
3151      3152       78      745         25725            1    credit_card   
1384      1385       78      867         25725            1           cash   
3085      3086       78      910         25725            1           cash   
2548      2549       78      861         25725            1           cash   
160        161       78      990         25725            1    credit_card   
4584      4585       78      997         25725            1           cash   
1419      1420       78      912         25725            1           cash   
3440      3441       78      982         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1364      1365       42      797          1760            5           cash   
1367      1368       42      926          1408            4           cash   
1471      1472       42      907          1408            4          debit   
3538      3539       43      830          1086            6          debit   

               created_at  

In [20]:
df_78 = sales_df.loc[sales_df['shop_id'] == 78]
df_78

order_id  shop_id  user_id  order_amount  total_items payment_method  \
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
1056      1057       78      800         25725            1          debit   
1193      1194       78      944         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1259      1260       78      775         77175            3    credit_card   
1384      1385       78      867         25725            1           cash   
1419      1420       78      912         25725            1           cash   
1452      1453       78      812         25725            1    credit_card   
1529      1530       78      810         51450            2           cash   
2270      2271       78      855         25725            1    credit_card   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2548      2549       78      861         25725            1           cash   
2564      2565       78      915         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2773      2774       78      890         25725            1           cash   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
2906      2907       78      817         77175            3          debit   
2922      2923       78      740         25725            1          debit   
3085      3086       78      910         25725            1           cash   
3101      3102       78      855         51450            2    credit_card   
3151      3152       78      745         25725            1    credit_card   
3167      3168       78      927         51450            2           cash   
3403      3404       78      928         77175            3          debit   
3440      3441       78      982         25725            1          debit   
3705      3706       78      828         51450            2    credit_card   
3724      3725       78      766         77175            3    credit_card   
3780      3781       78      889         25725            1           cash   
4040      4041       78      852         25725            1           cash   
4079      4080       78      946         51450            2           cash   
4192      4193       78      787         77175            3    credit_card   
4311      4312       78      960         51450            2          debit   
4412      4413       78      756         51450            2          debit   
4420      4421       78      969         77175            3          debit   
4505      4506       78      866         25725            1          debit   
4584      4585       78      997         25725            1           cash   
4715      4716       78      818         77175            3          debit   
4918      4919       78      823         25725            1           cash   

               created_at  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
511    2017-03-09 7:23:14  
617   2017-03-18 11:18:42  
691   2017-03-27 22:51:43  
1056  2017-03-15 10:16:45  
1193  2017-03-16 16:38:26  
1204  2017-03-17 22:32:21  
1259   2017-03-27 9:27:20  
1384  2017-03-17 16:38:06  
141

### The second type of problematic transaction is one where the number of items purchased and the order amount are significantly out of proportion. These transactions all occured at shop 78 and no transactions that appear reasonable happened at this shop. It would be helpful to be able to look into shop 78 but since this is not possible, the best option is to exclude all data from this shop in order to avoid skewing the results of AOV calculations. (Best guess here: either purchase data was entered incorrectly and 25725 for one item should be 257.25 for one item or this store was closing and these are the result of selling off fixtures and other assets.)

In [21]:
# exclude data from store 78
cleaned_df = step_one[step_one.shop_id != 78]
cleaned_df.head(50)


order_id  shop_id  user_id  order_amount  total_items payment_method  \
1364      1365       42      797          1760            5           cash   
1367      1368       42      926          1408            4           cash   
1471      1472       42      907          1408            4          debit   
3538      3539       43      830          1086            6          debit   
4141      4142       54      733          1064            8          debit   
3513      3514       42      726          1056            3          debit   
2987      2988       42      819          1056            3           cash   
938        939       42      808          1056            3    credit_card   
3077      3078       89      754           980            5          debit   
2494      2495       50      757           965            5          debit   
1563      1564       91      934           960            6          debit   
4847      4848       13      993           960            6           cash   
2307      2308       61      723           948            6    credit_card   
3532      3533       51      828           935            5           cash   
1256      1257        6      942           935            5    credit_card   
2560      2561        6      845           935            5    credit_card   
2039      2040       11      756           920            5    credit_card   
3073      3074       90      877           890            5          debit   
1150      1151       82      853           885            5          debit   
879        880       60      870           885            5          debit   
4523      4524       26      995           880            5    credit_card   
2032      2033       88      798           880            5           cash   
4952      4953       26      786           880            5           cash   
1946      1947       33      866           865            5           cash   
4958      4959       70      711           865            5    credit_card   
2353      2354       27      811           845            5           cash   
1962      1963       46      879           830            5          debit   
522        523       46      761           830            5    credit_card   
2967      2968       46      774           830            5          debit   
3865      3866       68      815           816            6          debit   
1123      1124       29      911           815            5    credit_card   
771        772       19      818           815            5          debit   
3927      3928       97      979           810            5    credit_card   
2757      2758       66      772           805            5    credit_card   
3438      3439       66      842           805            5    credit_card   
742        743       12      727           804            4           cash   
1764      1765       12      789           804            4          debit   
1923      1924       91      880           800            5    credit_card   
4619      4620       91      926           800            5    credit_card   
1628      1629       91      982           800            5          debit   
4870      4871       61      958           790            5    credit_card   
2136      2137       61      931           790            5          debit   
3252      3253       67      706           786            6    credit_card   
994        995       89      723           784            4           cash   
2140      2141       89      767           784            4           cash   
4295      4296       99      760           780            4           cash   
264        265       16      950           780            5    credit_card   
2670      2671       18      740           780            5          debit   
4711      4712       86      883           780            6           cash   
219        220       23      884           780            5    credit_card   

               created_at  